homeassistant/components/picnic/coordinator.py

```diff
@@ -0,0 +1,142 @@
+"""Coordinator to fetch data from the Picnic API."""
+import copy
+from datetime import timedelta
+import logging
+
+import async_timeout
+from python_picnic_api import PicnicAPI
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    ADDRESS,
+    SENSOR_CART_ITEMS_COUNT,
+    SENSOR_CART_TOTAL_PRICE,
+    SENSOR_COMPLETED_DELIVERIES,
+    SENSOR_LAST_ORDER_DELIVERY_TIME,
+    SENSOR_LAST_ORDER_ETA_END,
+    SENSOR_LAST_ORDER_ETA_START,
+    SENSOR_LAST_ORDER_SLOT_END,
+    SENSOR_LAST_ORDER_SLOT_START,
+    SENSOR_LAST_ORDER_STATUS,
+    SENSOR_LAST_ORDER_TOTAL_PRICE,
+    SENSOR_SELECTED_SLOT_END,
+    SENSOR_SELECTED_SLOT_MAX_ODER_TIME,
+    SENSOR_SELECTED_SLOT_MIN_ORDER_VALUE,
+    SENSOR_SELECTED_SLOT_START,
+    SENSOR_TOTAL_ORDERS,
+)
+
+
+class PicnicUpdateCoordinator(DataUpdateCoordinator):
+    """The coordinator to fetch data from the Picnic API at a set interval."""
+
+    def __init__(self, hass: HomeAssistant, picnic_api_client: PicnicAPI):
+        """Initialize the coordinator with the given Picnic API client."""
+        self.picnic_api_client = picnic_api_client
+
+        logger = logging.getLogger(__name__)
+        super().__init__(
+            hass,
+            logger,
+            name="Picnic coordinator",
+            update_interval=timedelta(minutes=30),
+            update_method=self.async_update_data,
+        )
+
+    async def async_update_data(self):
+        """Fetch data from API endpoint."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self.fetch_data)
+        except Exception as err:  # pylint: disable=broad-except
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    def fetch_data(self):
+        """Fetch the data from the Picnic API and return a flat dict with only needed sensor data."""
+        # Fetch from the API and pre-process the data
+        user = self.picnic_api_client.get_user()
```

Comment:
Do we always need the user first before fetching the cart ?

Reply:
Oh I see now that there are sensors that rely on user data and sensors that rely on cart data. It would make more sense to create multiple update coordinators. That way you won't be fetching user data the whole time if no user sensors are enabled.

Reply:
I've removed the sensors that rely on user information and made it such that the user endpoint only needs to be called once after initializing the coordinator.